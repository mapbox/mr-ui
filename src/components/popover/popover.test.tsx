import React, { useState } from 'react';
import Popover from './popover';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

const PopoverTest = (props) => {
  const [active, setActive] = useState(false);
  return (
    <Popover onExit={() => setActive(false)} active={active} {...props}>
      <button data-testid="trigger" onClick={() => setActive(!active)}>
        trigger
      </button>
    </Popover>
  );
};

describe('Popover', () => {
  const props = {
    content: 'Popup content'
  };

  describe('basic', () => {
    test('renders', async () => {
      const user = userEvent.setup();
      const { baseElement } = render(<PopoverTest {...props} />);
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('dark', () => {
    function renderContent() {
      return 'Content is rendered';
    }

    const props = {
      coloring: 'dark',
      content: renderContent
    };

    test('renders', async () => {
      const user = userEvent.setup();
      const { baseElement } = render(<PopoverTest {...props} />);
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('warning', () => {
    const props = {
      coloring: 'warning',
      content: <span>HI there</span>
    };

    test('renders', async () => {
      const user = userEvent.setup();
      const { baseElement } = render(<PopoverTest {...props} />);
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('no arrow', () => {
    const props = {
      hasPointer: false,
      content: <div>no arrow</div>
    };

    test('renders', async () => {
      const user = userEvent.setup();
      const { baseElement } = render(<PopoverTest {...props} />);
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
      expect(baseElement).toMatchSnapshot();
    });
  });

  describe('clickOutsideCloses', () => {
    test('does not close when clickOutsideCloses is false and user clicks somewhere else', async () => {
      const props = {
        clickOutsideCloses: false
      };

      const user = userEvent.setup();
      render(
        <div>
          <button data-testid="other">Click me</button>
          <PopoverTest {...props} />
        </div>
      );
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });

      await user.click(screen.getByTestId('other'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
    });

    test('closes when clickOutsideCloses is true and user clicks somewhere else', async () => {
      const props = {
        clickOutsideCloses: true
      };

      const user = userEvent.setup();
      render(
        <div>
          <button data-testid="other">Click me</button>
          <PopoverTest {...props} />
        </div>
      );
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });

      await user.click(screen.getByTestId('other'));
      await waitFor(() => {
        expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
      });
    });
  });

  describe('all options', () => {
    const props = {
      placement: 'left',
      padding: 'none',
      hasPointer: false,
      hideWhenAnchorIsOffscreen: true,
      allowPlacementAxisChange: false,
      containWithinViewport: false,
      clickOutsideCloses: false,
      escapeCloses: false,
      onExit: () => {},
      content: (
        <>
          <span className="txt-kbd">Greetings</span>
          <span className="color-green txt-kbd">human</span>
        </>
      ),
      passthroughProps: {
        'data-testid': 'foo'
      }
    };

    test('renders', async () => {
      const user = userEvent.setup();
      const { baseElement } = render(<PopoverTest {...props} />);
      await user.click(screen.getByTestId('trigger'));
      await waitFor(() => {
        expect(screen.getByRole('dialog')).toBeInTheDocument();
      });
      expect(baseElement).toMatchSnapshot();
    });
  });
});
