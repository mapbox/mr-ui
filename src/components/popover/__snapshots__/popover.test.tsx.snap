// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover all options renders 1`] = `
<body>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <span
      style="display: contents;"
    >
      <button
        aria-controls="radix-:r6:"
        aria-expanded="true"
        aria-haspopup="dialog"
        data-state="open"
        data-testid="trigger"
        type="button"
      >
        trigger
      </button>
    </span>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate(-6px, 0px); min-width: max-content; z-index: auto; --radix-popper-available-width: -6px; --radix-popper-available-height: 0px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 0px 50%;"
  >
    <div
      class="bg-white border--white color-text border shadow-darken25 round border border-1 border--border--white"
      data-align="center"
      data-side="left"
      data-state="open"
      data-testid="foo"
      id="radix-:r6:"
      role="dialog"
      style="z-index: auto; --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height); opacity: 0;"
      tabindex="-1"
    >
      <span
        class="txt-kbd"
      >
        Greetings
      </span>
      <span
        class="color-green txt-kbd"
      >
        human
      </span>
    </div>
  </div>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`Popover basic renders 1`] = `
<body>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <span
      style="display: contents;"
    >
      <button
        aria-controls="radix-:r0:"
        aria-expanded="true"
        aria-haspopup="dialog"
        data-state="open"
        data-testid="trigger"
        type="button"
      >
        trigger
      </button>
    </span>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate(6px, 0px); min-width: max-content; z-index: auto; --radix-popper-available-width: -6px; --radix-popper-available-height: 0px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 0px 0px;"
  >
    <div
      class="bg-white border--white color-text border shadow-darken25 round px12 py12 border border-1 border--border--white"
      data-align="center"
      data-side="right"
      data-state="open"
      id="radix-:r0:"
      role="dialog"
      style="z-index: auto; --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height);"
      tabindex="-1"
    >
      Popup content
      <span
        style="position: absolute; left: 0px; transform-origin: 0 0; transform: translateY(50%) rotate(90deg) translateX(-50%); top: 0px;"
      >
        <svg
          fill="#fff"
          height="6"
          offset="6"
          preserveAspectRatio="none"
          style="display: block;"
          viewBox="0 0 30 10"
          width="12"
        >
          <polygon
            points="0,0 30,0 15,10"
          />
        </svg>
      </span>
    </div>
  </div>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`Popover dark renders 1`] = `
<body>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <span
      style="display: contents;"
    >
      <button
        aria-controls="radix-:r1:"
        aria-expanded="true"
        aria-haspopup="dialog"
        data-state="open"
        data-testid="trigger"
        type="button"
      >
        trigger
      </button>
    </span>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate(6px, 0px); min-width: max-content; z-index: auto; --radix-popper-available-width: -6px; --radix-popper-available-height: 0px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 0px 0px;"
  >
    <div
      class="bg-gray-dark border--gray-dark color-white border shadow-darken25 round px12 py12 border border-1 border--border--gray-dark"
      data-align="center"
      data-side="right"
      data-state="open"
      id="radix-:r1:"
      role="dialog"
      style="z-index: auto; --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height);"
      tabindex="-1"
    >
      Content is rendered
      <span
        style="position: absolute; left: 0px; transform-origin: 0 0; transform: translateY(50%) rotate(90deg) translateX(-50%); top: 0px;"
      >
        <svg
          fill="#0E2127"
          height="6"
          offset="6"
          preserveAspectRatio="none"
          style="display: block;"
          viewBox="0 0 30 10"
          width="12"
        >
          <polygon
            points="0,0 30,0 15,10"
          />
        </svg>
      </span>
    </div>
  </div>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`Popover no arrow renders 1`] = `
<body>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <span
      style="display: contents;"
    >
      <button
        aria-controls="radix-:r3:"
        aria-expanded="true"
        aria-haspopup="dialog"
        data-state="open"
        data-testid="trigger"
        type="button"
      >
        trigger
      </button>
    </span>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate(6px, 0px); min-width: max-content; z-index: auto; --radix-popper-available-width: -6px; --radix-popper-available-height: 0px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 0px 50%;"
  >
    <div
      class="bg-white border--white color-text border shadow-darken25 round px12 py12 border border-1 border--border--white"
      data-align="center"
      data-side="right"
      data-state="open"
      id="radix-:r3:"
      role="dialog"
      style="z-index: auto; --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height);"
      tabindex="-1"
    >
      <div>
        no arrow
      </div>
    </div>
  </div>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;

exports[`Popover warning renders 1`] = `
<body>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
  <div>
    <span
      style="display: contents;"
    >
      <button
        aria-controls="radix-:r2:"
        aria-expanded="true"
        aria-haspopup="dialog"
        data-state="open"
        data-testid="trigger"
        type="button"
      >
        trigger
      </button>
    </span>
  </div>
  <div
    data-radix-popper-content-wrapper=""
    style="position: fixed; left: 0px; top: 0px; transform: translate(6px, 0px); min-width: max-content; z-index: auto; --radix-popper-available-width: -6px; --radix-popper-available-height: 0px; --radix-popper-anchor-width: 0px; --radix-popper-anchor-height: 0px; --radix-popper-transform-origin: 0px 0px;"
  >
    <div
      class="bg-orange-faint border--orange-deep color-orange-deep border shadow-darken25 round px12 py12 border border-1 border--border--orange-deep"
      data-align="center"
      data-side="right"
      data-state="open"
      id="radix-:r2:"
      role="dialog"
      style="z-index: auto; --radix-popover-content-transform-origin: var(--radix-popper-transform-origin); --radix-popover-content-available-width: var(--radix-popper-available-width); --radix-popover-content-available-height: var(--radix-popper-available-height); --radix-popover-trigger-width: var(--radix-popper-anchor-width); --radix-popover-trigger-height: var(--radix-popper-anchor-height);"
      tabindex="-1"
    >
      <span>
        HI there
      </span>
      <span
        style="position: absolute; left: 0px; transform-origin: 0 0; transform: translateY(50%) rotate(90deg) translateX(-50%); top: 0px;"
      >
        <svg
          fill="#A55109"
          height="6"
          offset="6"
          preserveAspectRatio="none"
          style="display: block;"
          viewBox="0 0 30 10"
          width="12"
        >
          <polygon
            points="0,0 30,0 15,10"
          />
        </svg>
      </span>
    </div>
  </div>
  <span
    data-radix-focus-guard=""
    style="outline: none; opacity: 0; position: fixed; pointer-events: none;"
    tabindex="0"
  />
</body>
`;
